packages/gatsby-transformer-video/src/gatsby-node.js

```diff
@@ -0,0 +1,273 @@
+import { resolve, parse } from "path"
+
+import { GraphQLString, GraphQLInt, GraphQLFloat } from "gatsby/graphql"
+import { ensureDir } from "fs-extra"
+
+import FFMPEG from "./ffmpeg"
+
+class WrongFileTypeError extends Error {}
+
+const DEFAULT_ARGS = {
+  maxWidth: { type: GraphQLInt, defaultValue: 1920 },
+  maxHeight: { type: GraphQLInt, defaultValue: null },
+  duration: { type: GraphQLInt, defaultValue: null },
+  fps: { type: GraphQLInt, defaultValue: null },
+  saturation: { type: GraphQLFloat, defaultValue: 1 },
+  overlay: { type: GraphQLString, defaultValue: null },
+  overlayX: { type: GraphQLString, defaultValue: `center` },
+  overlayY: { type: GraphQLString, defaultValue: `center` },
+  overlayPadding: { type: GraphQLInt, defaultValue: 10 },
+  publicPath: {
+    type: GraphQLString,
+    defaultValue: `assets/videos`,
+  },
+  screenshots: { type: GraphQLString },
+  screenshotWidth: { type: GraphQLInt, defaultValue: 600 },
+}
+
+exports.createSchemaCustomization = ({ actions, schema }) => {
+  const { createTypes } = actions
+
+  const typeDefs = [
+    schema.buildObjectType({
+      name: `GatsbyVideo`,
+      fields: {
+        path: GraphQLString,
+        absolutePath: GraphQLString,
+        name: GraphQLString,
+        ext: GraphQLString,
+        codec: GraphQLString,
+        formatName: GraphQLString,
+        formatLongName: GraphQLString,
+        startTime: GraphQLFloat,
+        duration: GraphQLFloat,
+        size: GraphQLInt,
+        bitRate: GraphQLInt,
+        screenshots: `[GatsbyVideoScreenshot]`,
+      },
+    }),
+    schema.buildObjectType({
+      name: `GatsbyVideoScreenshot`,
+      fields: {
+        path: GraphQLString,
+        absolutePath: GraphQLString,
+      },
+    }),
+  ]
+
+  createTypes(typeDefs)
+}
+
+exports.createResolvers = async (
+  { createResolvers, store, cache, createNodeId, actions },
+  { ffmpegPath, ffprobePath, profiles = {} }
+) => {
+  const { createNode } = actions
+
+  const program = store.getState().program
+  const rootDir = program.directory
+  const cacheDir = resolve(
+    rootDir,
+    `node_modules`,
+    `.cache`,
+    `gatsby-transformer-video`
```

Comment:
> cache in node_modules as it is more persistent

but then `gatsby clean` don't clean up the user data cache - i guess that behaviour is surprising for the user

Reply:
Usually, yes! We had this discussion before with `gatsby-transformer-sqip`.

Let me explain my decision:

Transforming videos is slow, when I keep the cache in gatsbys `.cache`, an update to any dependency, any change to gatsby-config.js or running a production build after using develop mode will trigger new videos to be build.

This means in my test case:
* multiple times of waiting 15 minutes when developing the project, even when not touching video related stuff
* rerendering of the exact same videos when doing a test production build on my machine
* any case cleaning gatsby cache will trigger 1h of waiting on the CI as well. Means every new pull request, most updates to pull requests etc. trigger full video conversion

Keeping the number of reconversions of videos as low as possible should be our goal to achieve neat DX/UX and reducing the load on CI providers.

On the other hand, the `node_modules/.cache` will persist way better. CI environments like netlify & gatsby cloud (just tested it with latest version) keep this directory till you install new dependencies. Also, pull requests will receive the masters `node_modules/.cache` directory.


The downsides of the current implementation I see so far:

* When query field arguments get changed, the videos option hash will change and a new file is generated. But old files won't be deleted. (We could create a list of "active" video files and remove old files from the cache in a post build gatsby event)
* Installing dependencies is a case where it is very unlikely that the converted videos will result in another file as before. At least his is a more rare case.


I am very open for suggestions and gatsby-wide solutions. The issue is very much the same for https://github.com/gatsbyjs/gatsby/tree/gatsby-transformer-video :)

Reply:
To work around this issue, I came up with this. Still works great for one of my projects:

https://github.com/axe312ger/gatsby-plugin-sftp-cache

Basically uses a ftp server as cache. The great benefit here is, that I can do heavy processing of  files on my local development machine, sync the results via `gatsby build` to my FTP and my CI just has to download if they are missing files.

Reply:
i understand why you do this - i see it also with the thumbnails creation, for development they cleaned away ....

i think it should be solved gatsby-wide

is there an already existing issue for this?

Reply:
if you place it into `node_modules` then there should be a note in the readme

Reply:
There is no open issue for this. I know I discussed it in person with @KyleAMathews long time ago when working on the first version of SQIP

Also @pieh was involved back then.

Here are related comments I found:

https://github.com/gatsbyjs/gatsby/pull/5862 (this made the netlify cache plugin possible)
https://github.com/gatsbyjs/gatsby/issues/5224#issuecomment-387327378 (Details of the netlify workaround)


Ahh here, This mentions permanent cache and says its a v3 topic and on Kyles list: https://github.com/gatsbyjs/gatsby/pull/4205#issuecomment-368640240

Gatsby Cloud was also mentioned by Kyle at some point that it might help out at some point.